### DISCUSSION TITLE
Batching behavior

### DISCUSSION THREAD

**Discussion by jobrachem** [katzfuss-group/batram#31] [General]
This issue serves to document the behavior of `SimpleTM.fit(batch_size=...)` depending on batch size.

## Initial observation

In my work, I observed that performance as measured by log scores drastically deteriorate when using batching, compared to full-batch optimization:

<img width="593" height="367" alt="Image" src="https://github.com/user-attachments/assets/978f1534-ae46-4c28-a4c3-331252e81b9d" />

Some details:

- The dataset is the global log precipitation rate data used in the JASA paper (https://github.com/katzfuss-group/BaTraMaSpa/tree/main/data), has about 55k locations and 98 replications of the field.
- Optimizer is rmsprop,
- I'm using a learning rate scheduler.
- I'm using early stopping based on four validation fields.

Relevant code excerpt:

```python
tm = SimpleTM(
    train_data,
    theta_init=None,
    linear=False,
    smooth=1.5,
    nug_mult=4.0,
)

stopper = EarlyStopper(
    patience=10, min_diff=0.3
)

opt = torch.optim.RMSprop(tm.parameters(), lr=0.01)
sched = torch.optim.lr_scheduler.CosineAnnealingLR(opt, 1000)

start = time.time()
res = tm.fit(
    num_iter=1000,
    init_lr=0.01,
    test_data=validate_data, # shape (4, nloc), i.e. 4 validation replications not used in training
    optimizer=opt,
    stopper=stopper,
    batch_size=1024,
    scheduler=sched,
)
```

Code and data for batching experimentation can be downloaded here: https://owncloud.gwdg.de/index.php/s/lQjYTIHEy9jtWfA

## Learning rate experiment

- An initial learning rate of 0.01 could simply to be too high in this case.
- I am running an experiment with different learning rates
- Not all runs have finished, but the initial results show that the learning rate indeed makes a large difference.
- It seems that the predictive performance issue is a learning rate / hyperparameter tuning issue.

<img width="722" height="491" alt="Image" src="https://github.com/user-attachments/assets/355808f9-c1cc-4c10-82ba-c53ce901c633" />

## Batching does not seem to help with memory usage

I ran just two iterations of the fitting loop with all locations and 80 training replications and captured memory usage using `/usr/bin/time` (MacOS).

These are the results:

```
276.79 real       123.29 user       235.85 sys
         23970693120  maximum resident set size
                   0  average shared memory size
                   0  average unshared data size
                   0  average unshared stack size
            27941540  page reclaims
                 305  page faults
                   0  swaps
                   0  block input operations
                   0  block output operations
                   0  messages sent
                   0  messages received
                   0  signals received
             1535508  voluntary context switches
             8322066  involuntary context switches
       1500758364525  instructions retired
       1059664334562  cycles elapsed
         43078241024  peak memory footprint
```

- Maximum resident set size: ~22 GB
- Peak memory footprint: ~40 GB

I would expect batched optimization to use much less memory.

**Comment by danjdrennan** [comment]
Thanks for sharing, especially the memory footprint (I agree this is unexpected
and strange).

For batch sizes, my understanding of SGD-like algorithms is that smaller batch
sizes will lead to higher variance gradient estimators. It may be possible to
correct for that with some adjustment, but we have not done so up to this point.
I expect this is an easy fix compared to the memory issues you mentioned.

My initial guess about memory utilization is that garbage collection doesn't
happen at the end of each iteration, and therefore you have several expired
buffers of data in flight simultaneously, coupled with additional memory
overhead for sampling the minibatches. Some thoughts:

- We can probably monitor when GC triggers to determine how much memory we
  expect to use at any given time
- We may be able to control the GC cycles to trigger at the end of each epoch so
  that we are not holding expired/unreachable buffers beyond a certain point
- It may be possible to reuse one or two minibatch buffers to avoid GC cycles
  entirely while also reducing memory consumption
